geode-core/src/test/java/org/apache/geode/management/internal/cli/commands/CreateRegionCommandTest.java

```diff
@@ -189,29 +191,124 @@ public void defaultValues() throws Exception {
+    ClassName writer = (ClassName) result.getParamValue("cache-writer");
+    assertThat(writer.getClassName()).isEqualTo("my.abc");
+    assertThat(writer.getInitProperties()).isNotNull().isEmpty();
+  }
+
+  @Test
+  public void parseDeclarableWithClassAndProps() {
+    String json = "{'k1':'v1','k2':'v2'}";
+    GfshParseResult result =
+        parser.parse("create region --name=region --cache-writer=my.abc?" + json);
+    ClassName writer = (ClassName) result.getParamValue("cache-writer");
+    assertThat(writer.getClassName()).isEqualTo("my.abc");
+    assertThat(writer.getInitProperties()).containsKeys("k1", "k2");
+  }
+
+  @Test
+  public void parseDeclarableWithJsonWithSpace() {
+    String json = "{'k1' : 'v   1', 'k2' : 'v2'}";
+    GfshParseResult result =
+        parser.parse("create region --name=region --cache-writer=\"my.abc?" + json + "\"");
+    ClassName writer = (ClassName) result.getParamValue("cache-writer");
+    assertThat(writer.getClassName()).isEqualTo("my.abc");
+    assertThat(writer.getInitProperties()).containsOnlyKeys("k1", "k2").containsEntry("k1",
+        "v   1");
+  }
+
+  @Test
+  public void cacheListenerClassOnly() {
+    GfshParseResult result =
+        parser.parse("create region --name=region --cache-listener=my.abc,my.def");
+    ClassName[] listeners = (ClassName[]) result.getParamValue("cache-listener");
+    assertThat(listeners).hasSize(2).contains(new ClassName("my.abc"), new ClassName("my.def"));
+  }
+
+  @Test
+  public void cacheListenerClassAndProps() {
+    String json1 = "{'k1':'v1'}";
+    String json2 = "{'k2':'v2'}";
+    GfshParseResult result = parser
+        .parse("create region --name=region --cache-listener=my.abc?" + json1 + ",my.def?" + json2);
+    ClassName[] listeners = (ClassName[]) result.getParamValue("cache-listener");
+    assertThat(listeners).hasSize(2).contains(new ClassName("my.abc", json1),
+        new ClassName("my.def", json2));
+  }
+
+  @Test
+  public void cacheListenerClassAndJsonWithComma() {
+    String json1 = "{'k1':'v1','k2':'v2'}";
+    String json2 = "{'k2':'v2'}";
+    GfshParseResult result = parser
+        .parse("create region --name=region --cache-listener=my.abc?" + json1 + ",my.def?" + json2);
+    ClassName[] listeners = (ClassName[]) result.getParamValue("cache-listener");
+    assertThat(listeners).hasSize(2).contains(new ClassName("my.abc", json1),
+        new ClassName("my.def", json2));
+  }
+
+  @Test
+  public void cacheListenerClassAndJsonWithCommaAndSpace() {
+    String json1 = "{'k1' : 'v1', 'k2' : 'v2'}";
+    String json2 = "{'k2' : 'v2'}";
+    GfshParseResult result = parser.parse("create region --name=region --cache-listener=\"my.abc?"
+        + json1 + ",my.def?" + json2 + "\"");
+    ClassName[] listeners = (ClassName[]) result.getParamValue("cache-listener");
+    assertThat(listeners).hasSize(2).contains(new ClassName("my.abc", json1),
+        new ClassName("my.def", json2));
   }
 
```

Comment:
Please add a test with invalid JSON - I don't think it's handling that gracefully.